src/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs

```diff
@@ -667,7 +667,7 @@ private T CreateCustomNoMatchDelegate(MethodInfo invoke)
                 Expression.Block(
                     Expression.Call(
                         typeof(CallSiteOps).GetMethod(nameof(CallSiteOps.SetNotMatched)),
-                        @params.First()
+                        @params[0]
```

Comment:
Getting methodinfo with no parameters would indicate a serious bug in CallSite related machinery or the compiler generated code (as opposed to API misuse). 
I do not think we care much about kind of exception that is thrown here.